### PR TITLE
Spring Data: Fix interface projection using Page and Slice

### LINKED ISSUE #21616: Spring data Page does not work with interface projection
### Describe the bug

When using `org.springframework.data.domain.Page` with an interface projection, we get an error.

The same query works fine when used with a `java.util.List` instead of a Page 

### Expected behavior

A Page of projected entities

### Actual behavior

We get an exception :

```
java.lang.ClassCastException: class [Ljava.lang.Object; cannot be cast to class <redacted1>.$SmallItem ([Ljava.lang.Object; is in module java.base of loader 'bootstrap'; <redacted1>.$SmallItem  is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @1c91a09d)
```

### How to Reproduce?

Create a repository with returning a `org.springframework.data.domain.Page` while using a projection to an interface.

```java
@Repository
public interface ItemRepository extends CrudRepository<Item, Long> {
    @Query(
        value = "select distinct(c.name) as name, c.id as id, c.version as version from Item c where lower(c.name) like :name" 
        ,countQuery = "select count(distinct(c.name)) from Item c where lower(c.name) like :name"
    )
    Page<SmallItem> findAllByName(@Param("name") String name, Pageable page);

    interface SmallItem {
        String getName();
        String getVersion();
        Long getId();
    }

}
```

### Output of `uname -a` or `ver`

Linux 5.4.0-47-generic #51-Ubuntu SMP  x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.11" 2021-04-20 OpenJDK Runtime Environment (build 11.0.11+9-Ubuntu-0ubuntu2.20.04) OpenJDK 64-Bit Server VM (build 11.0.11+9-Ubuntu-0ubuntu2.20.04, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.4.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.6.3 Maven home: /usr/share/maven Java version: 11.0.11, vendor: Ubuntu, runtime: /usr/lib/jvm/java-11-openjdk-amd64 Default locale: en_US, platform encoding: UTF-8 OS name: "linux", version: "5.4.0-47-generic", arch: "amd64", family: "unix"

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Sgitario**
The interface project worked fine using Lists, but not using Page and Slice. Example:

```java
@Repository
public interface BasicTypeDataRepository extends CrudRepository<BasicTypeData, Integer> {

    // works
    @Query(value = "select id, doubleValue from BasicTypeData")
    List<WithDoubleValue> findAllDoubleValuesToList();

   // it does not work (1)
    @Query(value = "select id, doubleValue from BasicTypeData")
    Page<WithDoubleValue> findAllDoubleValuesToPage();
}
```

This PR fixes the interface projection when returning a Page object (1).

Fixes https://github.com/quarkusio/quarkus/issues/21616

### CODE REVIEW

**Comment by famod** [comment]
I never worked with it, but what about `Slice`? `Page` extends `Slice`, btw...

**Comment by Sgitario** [comment]
> I never worked with it, but what about `Slice`? `Page` extends `Slice`, btw...

Well spotted! I've already fixed it for Slice too. Thanks.

File: extensions/spring-data-jpa/deployment/src/test/java/io/quarkus/spring/data/deployment/BasicTypeDataRepositoryTest.java [hunk 1]
```diff
@@ -75,6 +79,33 @@ public void testTimeZonesByLocale() {
...
+        List<WithDoubleValue> list = repo.findAllDoubleValuesToList();
+        Set<Double> actual = list.stream().map(WithDoubleValue::getDoubleValue).collect(Collectors.toSet());
+        assertThat(actual).isNotEmpty().contains(Math.PI);
+    }
+
+    @Test
+    @Order(6)
+    @Transactional
+    public void testMapInterfacesUsingPage() {
+        Page<WithDoubleValue> list = repo.findAllDoubleValuesToPage();
```
Comment (famod) [inline extensions/spring-data-jpa/deployment/src/test/java/io/quarkus/spring/data/deployment/BasicTypeDataRepositoryTest.java:95]: Nitpick: `list` variable name is a bit misleading. Same for the `Slice` test method.

File: extensions/spring-data-jpa/deployment/src/test/java/io/quarkus/spring/data/deployment/BasicTypeDataRepositoryTest.java [hunk 2]
```diff
@@ -75,6 +79,33 @@ public void testTimeZonesByLocale() {
...
+        Set<Double> actual = list.stream().map(WithDoubleValue::getDoubleValue).collect(Collectors.toSet());
+        assertThat(actual).isNotEmpty().contains(Math.PI);
+    }
```
Comment (famod) [inline extensions/spring-data-jpa/deployment/src/test/java/io/quarkus/spring/data/deployment/BasicTypeDataRepositoryTest.java:98]: What about testing `getTotalElements()` etc.? Same for the `Slice` test method.

**Review by famod — APPROVED** [review]
LGTM, only two small remarks/suggestions.

**Review by geoand — COMMENTED** [review]
Thanks!

@famod covered me on this :)

File: extensions/spring-data-jpa/deployment/src/test/java/io/quarkus/spring/data/deployment/BasicTypeDataRepositoryTest.java (hunk 2 above)
Comment (Sgitario) [reply extensions/spring-data-jpa/deployment/src/test/java/io/quarkus/spring/data/deployment/BasicTypeDataRepositoryTest.java:110]: Done

File: extensions/spring-data-jpa/deployment/src/test/java/io/quarkus/spring/data/deployment/BasicTypeDataRepositoryTest.java (hunk 1 above)
Comment (Sgitario) [reply extensions/spring-data-jpa/deployment/src/test/java/io/quarkus/spring/data/deployment/BasicTypeDataRepositoryTest.java]: Changed

**Comment by Sgitario** [comment]
Updated PR to cover some functionality around `Pageable`.

**Pull request merged**